### DISCUSSION TITLE
Library structure

### DISCUSSION THREAD

**Discussion by gnodar01** [CellProfiler/CellProfiler#4760] [Ideas]
I'm wondering if it makes sense to keep our current package structure for Library.

I'll use `expand_or_shrink_objects` as an example, but everything described applies more generally.

What we have currently:

```
library
├── __init__.py
├── functions
│  ├── __init__.py
│  ├── file_processing.py
│  ├── image_processing.py
│  └── object_processing.py
└── modules
   ├── __init__.py
   ├── _expand_or_shrink_objects.py
   ├── _module2.py
modules
├── __init__.py
├── expandorshrinkobjects.py
├── module2.py
```

with `library/modules/__init__.py`  having:

```
from ._expand_or_shrink_objects import expand_or_shrink_objects
from ._module2 import module2
```

and `modules/expandorshrinkobjects.py` having:

```
from cellprofiler.library.modules import expand_or_shrink_objects
```

Which works fine, and makes sense right now. We have a library module shadowing the GUI module, which imports it sort of as a singleton.

However, we often have an idiom where we control behaviour via some string flag. Like this in one in `modules/expandorshrinkobjects.py`:

```
from cellprofiler.library.modules import expand_or_shrink_objects

...
O_EXPAND = "Expand objects by a specified number of pixels"
O_EXPAND_INF = "Expand objects until touching"
...

class ExpandOrShrinkObjects(Module):
    def create_settings(self):
        # a bunch of options set here
        ...
    
    def do_labels(self):
        if self.operation == O_EXPAND:
            return expand_or_shrink_objects('expand_defined_pixels', labels, iterations=self.iterations.value)
        ...
        elif self.operation == O_EXPAND_INF:
            return expand_or_shrink_objects('expand_infinite', labels)
        ...
```

Passing in `"expand_defined_pixels"` or `"expand_infinite"`  is a bit brittle. So what I'd like to do is create an enum enforced by `library/modules/_expand_or_shrink_objects.py`:

```
from enum import StrEnum

class MODE(StrEnum):
    EXPAND = 'expand_defined_pixels'
    EXPAND_INF = 'expand_infinite'

def expand_or_shrink_objects(mode, labels, fill=None, iterations=None):
    if mode == MODE.EXPAND:
        return expand_defined_pixels(labels, iterations=iterations)
    elif mode == EXPAND_INF:
        return expand_until_touching(labels)
    ...
```

Allowing us to do this:

```
    def do_labels(self):
        if self.operation == O_EXPAND:
            return expand_or_shrink_objects(MODE.EXPAND, labels, iterations=self.iterations.value)
        ...
        elif self.operation == O_EXPAND_INF:
            return expand_or_shrink_objects(MODE.EXPAND_INF, labels)
```

The problem comes in here. There are currently two or three main ways to make this enum available in `modules/expandorshrinkobjects.py`.

The first is to import it directly:

```
from cellprofiler.library.modules import expand_or_shrink_objects
from cellprofiler.library.modules._expand_or_shrink_object import MODE
```

Which works, but now we have two lines of imports, and worse we've imported from a subpackage with a leading underscore in front of it that, by convention, marks it as "internal use only".

The second is to make it available through `library/modules/__init__.py`:

```
from ._expand_or_shrink_objects import expand_or_shrink_objects, MODE
```

Which mean we can have this in `modules/expandorshrinkobjects.py`:

```
from cellprofiler.library.modules import expand_or_shrink_objects, MODE
```

The problem with that is `MODE` is fairly generic. For instance `_module2.py` may also have a `MODE` enum, leading to:

```
from ._expand_or_shrink_objects import expand_or_shrink_objects, MODE as EXPANDORSHRINK_MODE
from ._module2 import module2, MODE as MODULE2_MODE
```

In `modules/expandorshrinkobjects.py`, we can now do:

```
from cellprofiler.library.modules import expand_or_shrink_objects, EXPANDORSHRINK_MODE
```

Which is slightly better, but it's verbose, icky, and also clutters the `cellprofiler.library.modules` namespace.

In terms of solutions, there are multiple approaches.

We could just do away with the leading underscore convention.

So in `modules/expandorshrinkobjects.py` we'd have:

```
from cellprofiler.library.modules.expand_or_shrink_objects import expand_or_shrink_objects, MODE
```

I kind of hate how long and verbose that is. However it does allow us to move away from the singleton model and have additional methods in `library/modules/_expand_or_shrink_objects.py`, importable in `modules/expandorshrinkobjects.py` as such:

```
from cellprofiler.library.modules.expand_or_shrink_objects import (
    expand_or_shrink_objects,
    MODE,
    some_helper_fn
)
```

Alternatively, we could decouple the library modules from library module options, utils, etc.

So in `modules/expandorshrinkobjects.py` we'd have:

```
from cellprofiler.library.modules import expand_or_shrink_objects
from cellprofiler.library.opts.expand_or_shrink_object import MODE
from cellprofiler.library.utils.expand_or_shrink_object import some_helper_fn
```

The difference of `library/utils` vs `library/functions`, is that `utils` are module specific, ie `library/utils/expand_or_shrink_object`. There are obvious downside here too, the main one being the first 2 imports are *always* together, since you need `MODE` to invoke `expand_or_shrink_object`. Depending on the contents of `some_helper_fn`, you might also need that import most/all of the time. Maintenance also becomes a bit burdensome, because now we have to track `expand_or_shrink_object` related things across many different files in different directories.

As another alternative, we could use [function attributes](https://peps.python.org/pep-0232/), so in `library/modules/_expand_or_shrink_objects.py`:

```
from enum import StrEnum

class MODE(StrEnum):
    EXPAND = 'expand_defined_pixels'
    EXPAND_INF = 'expand_infinite'

def expand_or_shrink_objects(mode, labels, fill=None, iterations=None):
    if mode == MODE.EXPAND:
        return expand_defined_pixels(labels, iterations=iterations)
    elif mode == EXPAND_INF:
        return expand_until_touching(labels)
    ...

def some_helper_fn():
    ...

expand_or_shrink_object.MODE = MODE
expand_or_shrink_object.some_helper_fn = some_helper_fn
```

In `modules/expandorshrinkobjects.py` we'd have:

```
from cellprofiler.library.modules import expand_or_shrink_objects

...
    def do_labels(self):
        if self.operation == O_EXPAND:
            return expand_or_shrink_objects(expand_or_shrink_objects.MODE.EXPAND, labels, iterations=self.iterations.value)
...

# do something with the helper fn
expand_or_shrink_objects.some_helper_fn()
```

I don't like this approach at all. Obviously it's super annoying to always have to spell out `expand_or_shrink_objects.MODE.EXPAND`, etc. It's also a **huge** headache to deal with for incorporate typing support.

There are other solutions, but I'd like to get thoughts first.

**Comment by bethac07** [comment]
Is the majority of your concern around that particular module? Because that was a super early test and there's a [PR open](https://github.com/CellProfiler/CellProfiler/pull/4682) to clean it up, just waiting for someone not-me to approve and merge it

**Reply by gnodar01** [reply]
No, it applies more generally. For instance `enhanceedges` dispatches on a `method` argument, which is a string and I would prefer it to be an enum.

**Reply by gnodar01** [reply]
We could also just not use Enums and stick to strings. I haven't (yet) found an example where module-specific helper functions or module-specific constants are necessary. If we expect they will be in the future however, we'll probably want to restructure somehow. If we don't expect it, using strings instead of enums for dispatching isn't that big of a deal.

**Reply by callum-jpg** [reply]
I had left PR #4682 open to wait and see if we decided to remove the underscores from the module name (ie. `expand_or_shrink_objects` becomes `expandorshrinkobjects`) so it was consistent with the GUI, mentioned here: #4747

**Reply by gnodar01** [reply]
I'm not quite ready to weigh in on the linked issue quite yet. I'll have to think about it a bit more.

Just in case there might be confusion though, regardless of if we go with snake case, or all one word, for file names, in this discussion I'm referring to the **leading** underscore on the file name specifically. In python-land a single leading underscore means "for internal use only".  A user doing `import cellprofiler.modules._somemodule` or `import cellprofiler.modules._some_module` would get equally nervous seeing that leading underscore before `some`.

In fact it's a bit more than just convention, since the python interpreter itself takes leading underscores into account. It does not import variables and functions with leading underscores by default; it has to explicitly be told to import it. For filenames with leading underscore, the python interpreter doesn't care, but the convention is still there.

**Comment by callum-jpg** [comment]
What is it that makes passing in "expand_defined_pixels" or "expand_infinite" brittle? 

In #4682 I believe that the constant should be directly passed to the library level module call (while also being converted by some utility function that removes spaces and hyphens). This is currently the approach used in watershed: #4665. This is similar in the threshold module (without space removal since it's not required), though there is some tricky bits here in translating "otsu" + "three classes" into "multiotsu".

What would a hypothetical module-specific helper function look like? Are the functions in `cellprofiler.library.functions.image_processing` technically helper functions for `library.modules` threshold?

If a helper function were to be used, would enums just ensure that the same "string" was being passed to these helper functions? Ie. the constants would not live in the GUI modules, but instead in library?

**Reply by gnodar01** [reply]
> What is it that makes passing in "expand_defined_pixels" or "expand_infinite" brittle?

1) On the usage end, A user could misspell the argument and not know it's misspelled until runtime. With an enum, your IDE will give all possible values the argument could take, since they are pre-defined by the enum. You can't really misspell an enum, because the IDE will provide you with a little red squiggly line underneath immediately. Static analysis tools will also complain even if your IDE doesn't.

2) On the implementation end, if we allow for strings, we must always check that the provided string is an acceptable value and raise an exception if it is not. For enums we don't strictly have to (although it's good practice to do so anyway).

3) On the maintenance end, if we decide to modify an existing argument name, and we change the enum, our IDE can automatically refactor all usages of the function with the original argument to the new one (right click, rename symbol). With strings we have to do it by hand.

![Screenshot 2023-03-31 at 3 39 30 PM](https://user-images.githubusercontent.com/8888732/229214029-4b27f1c5-2399-47e5-a285-975e1c450df7.png)


> In https://github.com/CellProfiler/CellProfiler/pull/4682 I believe that the constant should be directly passed to the library level module call

That's fair. In the example of https://github.com/CellProfiler/CellProfiler/pull/4682 we define the mapping:

```
library_mapping = {
    O_SHRINK_INF: 'shrink_to_point',
    ...
}
```

which (if we were to go with the enum approach) would become: 

```
library_mapping = {
    O_SHRINK_INF: MODE.SHRINK_TO_POINT,
    ...
}
```

lending itself to 1-3 above.

In watershed, if you decide to change `O_DISTANCE` to something like "dist." (contrived, I know), you would then have to change the type value in `object_processing.py` and the value in the `if` statements. You'd also have to change the type value in `_watershed.py`. 

It is very far from a big deal. It's just one extra little thing.

> What would a hypothetical module-specific helper function look like?

Not sure, for us especially. For users of library, I was just thinking if there's some common transformation or preprocessing step that's always done to the input data before calling `some_module`, then it can help to expose that as a helper function. If that's not something we should expect, or we want to make the user responsible for doing that themselves, that's fine.

**Reply by callum-jpg** [reply]
Point 1 alone sells the idea of enum to me. In the initial examples you gave, I liked the `from cellprofiler.library.opts.expand_or_shrink_object import MODE` version you proposed.

Though if we consider that the methods that library uses are just what the GUI uses but with underscores instead of spaces/hyphens, then should the enum values be generated from the GUI level constants? This way, the constant is defined once inside the GUI modules and then propagated into enums (somehow), rather than being defined twice (once for GUI, once for enum). Then, the GUI module constants can be passed to the library function call and converted to "library grammar" with some other function (ie one that removes spaces/hyphens).

**Reply by gnodar01** [reply]
> should the enum values be generated from the GUI level constants

At runtime? If I'm understanding correctly, GUI is downstream of Library, so if anything it would have to be the other way around. We can't let the GUI affect anything in Library.

I mean, we can, but we probably shouldn't.

I think the `library_mapping` approach is quite nice. For watershed:

```cellprofiler/modules/watershed
from cellprofiler.library.opts.watershed import METHOD

...

library_mapping = {
    O_DISTANCE: METHOD.Distance
    O_MARKERS: "METHOD.Markers
    O_LOCAL: METHOD.Local
    O_REGIONAL: METHOD.Regional
    O_SHAPE: METHOD.Shape
    O_INTENSITY: METHOD.Intensity
}

...
def run(self):
    ...
        y_data, seeds = watershed(
                input_image=x_data,
                mask=mask_data,
                watershed_method=library_mapping[self.watershed_method.value],
               ...
                )
    ...
...
```

Which is pretty good for us, since we can let constants like `O_DISTANCE` take on any amount of spaces or hyphens, and we don't have to worry about converting them since they're agnostic to the values of `METHOD.Distance` and the rest. The enums could be integers, and still work just as well.

**Reply by bethac07** [reply]
Having library have an opts, and then the module-modules having a library_mapping works for me! I prefer text to random ints because that will be more annoying for tracing purposes if we accidentally cross-wire something.

**Reply by callum-jpg** [reply]
I see exactly what you mean - I think I was fixated on having it defined only once but that doesn't make sense dependency-wise for library.

With this in mind, I like the library_mapping strategy too - it's clear and concise.

**Comment by callum-jpg** [comment]
I was thinking of general helper functions that some modules may require. As far as I understand, the objects in CellProfiler are stored sequentially (ie. [1, 2, 3, 4], in part [see here](https://github.com/CellProfiler/core/blob/83506a0c332fa5a350237d550727eace408454ec/cellprofiler_core/object/_objects.py#L121)) and functionality relies on this. However, what if a user passes in a non-sequential label matrix? This could lead to unexpected results/errors. So, a helper function could make sure input labels are sequential, eg:
```
def make_consecutive(array):
    """Make a non-contiguous label matrix contiguous"""
    unique = np.unique(array)
    for new_val, unq in enumerate(unique):
        array[array == unq] = new_val
    return array
```

Alternatively, we could just return an error if the labels are not consecutive and direct users to make them consecutive themselves with some helper function library provides. Hopefully this would allow users to inspect that the labels had changed in a way they expect prior to the next steps in analysis.

**Reply by gnodar01** [reply]
I think the error is the right approach. It signals to the user they've done something wrong up to that point. We have no way of knowing exactly what it is they've done wrong, and therefore whether simply renumbering their labels for them is a sufficient fix. It should be up to them to figure out what went wrong and manually invoke the helper if it is indeed... helpful.

I think for your function, we want the least element in the list to be `1`, correct? In `make_consecutive` it's `0` as you have it, so you'd need `return array + 1`. You may also want to wrap `array` with `np.arrray(array)`, just in case they pass a standard python list.

**Reply by callum-jpg** [reply]
I think it's important to have the least element as 0 since that would represent the background, which we would want to maintain. 

The line I linked to above gets the indices from the ijv, which only stores labels and not background. There'd have to be additional functionality for making ijv consecutive.

**Reply by bethac07** [reply]
https://scikit-image.org/docs/stable/api/skimage.segmentation.html#skimage.segmentation.relabel_sequential might be helpful to use instead!

**Reply by sspathak** [reply]
A similar CP specific behavior is of type coercion.
The `Image` class of cellprofiler has a param `convert`  that tries to "coerce whatever dtype passed (other than bool or float) to a scaled image."
`frontend/convertobjectstoimage` forces this option to `True` when creating the output image.
But, the `Image` class won't be a part of `library` as it is a part of `core` 
Should I implement this behavior of type coercion in `library/functions/convertobjectstoimage` or `library/modules/convertobjectstoimage` or raise an exception instead?

**Comment by bethac07** [comment]
Excited to see this in #4758!

A couple q's, now that I have a more concrete example in mind - I'm going to make a separate comment for each, so we can thread them independently. I apologize for not getting my brain fully wrapped around these earlier, it's too full these days!
- If in general these are going to be called at the `library/function` level, does it make sense to structure them similarly - aka just have `opts/image_processing`, `opts/object_processing`, and `opts/measurement` ? I could go either way, because then it's more annoying at the module level, but I ultimately think it's less annoying to have to for each module remember what category it is than for in the library functions to have to know the name of the module, especially for library users who aren't already CellProfiler users.

**Reply by gnodar01** [reply]
I think that's the better approach. One thing that tripped me up and blocked me from realizing it's better at that level is that the values of the enum are the literal strings displayed in the module's corresponding setting in the GUI. That lead my brain to come up with "if there's an enumerable setting (eg dropdown) in Module X, then there must naturally be a corresponding `enumeration` in `cellprofiler.library.opts.module_x`.

If the enum reaches `library/function`, then it does make sense to have it live in `opts/[[image | object]_processing | measurement]`. One caveat to that though, is now `library/function` has a light "implicit dependency" on what things look like in the GUI, similar to `library/modules`, in that it is important for enum member values to be something legible and understandable in the GUI.

**Reply by bethac07** [reply]
> One caveat to that though, is now library/function has a light "implicit dependency" on what things look like in the GUI, similar to library/modules, in that it is important for enum member values to be something legible and understandable in the GUI.

Sure, but IIUC that was going to be true whether `library.opts` stored things as `opts/measurement` or `opts/measureimageoverlap` - ultimately, if we're pulling all the setting values (at least all the string ones) out of the modules, that's the caveat we're adding.

**Reply by gnodar01** [reply]
Yeah, fair enough. I am now thinking about modules like `expandorshrinkobjects`, where the `mode` parameter in the library function currently takes on values like `"expand_defined_pixels"`, and `"expand_infinite"`. When switching to enum, from the library function perspective it will just be getting the enum name `O_EXPAND` or `O_EXPAND_BY_MEASUREMENT` as an argument, but in the enum definition, the enum member values will now be the literals `"'Expand objects by a specified number of pixels"` or `"Expand objects by a specified number of pixels"`.

Also now realizing that here, the enum for the `mode` argument is only to control the behavior of `library/modules/_expandorshrinkobjects`, and never actually makes it to `library/functions` as an argument.

**Reply by bethac07** [reply]
<sarcasm voice, in case it doesn't convey over text well> I'm sorry, were you expecting CellProfiler to be consistent in how it does things?</sarcasm voice>

**Comment by bethac07** [comment]
- How would this work at the level of documentation? I'm thinking of if we had a sphinx/readthedocs guide to library, by no longer having the options for each function outlined IN the function, are those auto-documentation tools smart enough to pull the enum keys in? I can definitely see this majorly hurting usability if I'm a new user and I want to pull X function out of the library but it has one or more string args, and now I have to go look up the permissible values for those.

**Reply by gnodar01** [reply]
🤔

We'll have to figure out if something like [sphinx-autodoc-typehints](https://github.com/tox-dev/sphinx-autodoc-typehints) automatically infers the type of an enum to not just be the enum name, but the module where it's defined.

I would say "we can't be the first ones to want this", but I've been burned by that too much before.

**Reply by bethac07** [reply]
>I would say "we can't be the first ones to want this", but I've been burned by that too much before.

Yeah, we tend to be pretty far out on the bell curve of people trying to make code beginner-friendly.

**Reply by bethac07** [reply]
So it doesn't look fully automatic, but it also looks like it's a solved problem and that the solution is pretty generalizable

https://docs.bokeh.org/en/latest/docs/reference/sphinxext.html#module-bokeh.sphinxext.bokeh_enum

Relies on this
https://github.com/bokeh/bokeh/blob/72544d97b54b8ca3308473ceba235e225d115c87/src/bokeh/sphinxext/bokeh_enum.py
Which calls this
https://github.com/bokeh/bokeh/blob/72544d97b54b8ca3308473ceba235e225d115c87/src/bokeh/sphinxext/templates.py#L76
Which calls this, which sets up the template style
https://github.com/bokeh/bokeh/blob/72544d97b54b8ca3308473ceba235e225d115c87/src/bokeh/sphinxext/_templates/enum_detail.rst

**Comment by bethac07** [comment]
- Just checking - for modules we otherwise don't intend to put in the library, we aren't going to do this, right? What about the corner case where some module options we plan to put in the library, and some we don't? (At least one does exist, though of course it's skippable)

**Reply by gnodar01** [reply]
Yeah, I think it only makes sense to define an enum for something that actually makes it into library, whether a whole module or just an option.

One of the problems solved by the use of enums is that library doesn't have access to the top level constants in `cellprofiler`, and therefore we had the value of an argument defined in two places, in `cellprofiler` and again in the `if/else` statements of `library` (slightly mutated after conversion to remove hyphens/spaces). Enums make it so the actual string value is ever defined once between `library` and `cellprofiler`. If a module or option doesn't live in `library` then it's only defined once as a top level constant.

**Comment by bethac07** [comment]
This isn't related to 4758 per se but just writing it down here after I raised it in checkin yesterday because stuff leaks out of my brain a lot these days - how far do we take our ability to protect users from bad settings into the library? Lots of CellProfiler numerical settings have mins and/or maxes, are we going to add checks for all of them in the library? It would certainly be a nice thing to do but it's going to add complexity and time to do this.

**Reply by gnodar01** [reply]
For an argument who's type is `int`, but should be more restricted to something simple like "-1, 0, or 1", we can do this at the type level with `Literal[-1, 0, 1]`. For `int` types with lots of values (e.g. 0-100) or floats we'll have to do a runtime validity check.

**Reply by bethac07** [reply]
Yeah, there will be essentially 0 times when we can check it at the type level, I was (without being clear) talking about runtime validity checking - it would be really nice of us to take it on but there's a LOT of settings in CellProfiler, it might be a really big task to take on.

**Comment by bethac07** [comment]
This is library-structure-ish; how tight do we want to be with dependencies in library? I was surprised to realize matplotlib is a library dependency - for these lines of code, which are just to define the colors to pass into skimage. It's _nice_ to have this, but is it worth having mpl as a dependency if that's the only thing it does? It may NOT be the only thing we want it for, so fine to punt on this and let it settle, but if we want library to be relatively slim, we may have this and a few other corner cases to individually weigh

```
def overlay_objects(image, labels, opacity=0.3, max_label=None, seed=None, colormap="jet"):
    cmap = matplotlib.cm.ScalarMappable(cmap=matplotlib.cm.get_cmap(colormap))

    colors = cmap.to_rgba(
        numpy.arange(labels.max() if max_label is None else max_label)
    )[:, :3]
```

**Reply by gnodar01** [reply]
Yeah, we should move it out. I do remember including that as part of the post-monorepo refactor of library, but at the time I incorrectly assumed that matplotlib was already a dependency of scipy and/or scikit-image, so it wouldn't affect the package size. Turns out it's only an optional dependency in both, so removing matplotlib from our dependency list is the way to go.

**Reply by bethac07** [reply]
@sspathak re: ObjectsToImage

**Comment by sspathak** [comment]
Question about the role of `library/module` vs `library/functions`:
Context:
A number of modules that work on `Objects` use helper functions from `cellprofiler_library.functions.segmentation` to access various segmentation formats like sparse, dense, or derived properties such as the shape or indices in a segmentation. Core often caches the results of these functions:
```python
def get_dense(self):
        ...
        if self.__dense is not None:
            return self.__dense, self.__indices
        ...

        return self.__convert_sparse_to_dense()  # self.__dense gets updated in this call
```
In CellProfiler Library, we have some modules that could take advantage of these cached values to perform measurements and other actions without double computation.

Question: We have the following two options when assign responsibilities to `library/module` and `library/functions` -
1. `library/module` accepts a large set of inputs, including any cached values the user wishes to pass. It is responsible for computing the values in case the cache is not present, and finally it calls `library/functions` with the right arguments. `library/functions` expects to receive only those datatypes that it uses in its algorithm and is not responsible for performing any kind of conversions. As a result, the function signature in `library/functions` is minimal.
2. Both `library/module` and `library/functions` accept a large set of inputs. `library/module` merely performs type checking and ensures that all the parameters necessary for computation exist and are not `None` and forwards them to `library/functions`. If conversion between datatypes is needed, `library/functions` determines the optimal way to compute the derived properties before performing the actual computation.

I am leaning towards option 1 so that the function signatures in `library/functions` are short and minimal. Furthermore, it separates the logic needed to compute the derived properties and the functional code of the module's math. I'm open to option 2 in case there might be a benefit with that that I missed (and someone else feels strongly about).

**Reply by bethac07** [reply]
Can you give a specific example? My initial read is that those sorts of caching don't belong in library at all (since that should ultimately be a property of the CellProfiler object class, not anything that's CellProfiler-agnostic), but a specific use case would help. TY!

**Reply by gnodar01** [reply]
Agree with @bethac07 on the caching.

For 1)

>  It is responsible for computing the values in case the cache is not present

No, library is never responsible for transforming user input in a "smart way". It declares, via its arguments and their types, exactly what it wants, and errors if it doesn't get it. It can *reorganize* things (e.g. unpack a dictionary), and do some mild *conversions* (e.g. turn a tuple into a list or vice versa), but not transform them (e.g. turn a dense representation into a sparse one or vice versa). I'm even hesitant, though not unwilling, to allow for mild reshaping like turning a nxm numpy array into an mxn numpy array (technically not a transform since the underlying data buffer is untouched).  

For 2)

> Both library/module and library/functions accept a large set of inputs

`library/functions` certainly should not have a large set of inputs. `library/module` arguably shouldn't either. I really don't like option creep, and don't see it as necessary here. We don't need library to be this smart, and even if we did, a better design would be to introduce a new middleware, provided by library, to do these things.  I don't think that's high priority though. Right now, I fail to see how the classes in the frontend are not up to the task of handling caching. 

More broadly (outside of caching transformations of input values), I do see some appeal. Especially when we have relatively simple derived values.

If for example we have a `library/functions` function that takes in an argument which is derivable from some combination of its inputs, the temptation is to let the associated `library/module` function either take it in as an argument or compute it automatically. I think the overall approach there might be just to ask ourselves "is this something we can meaningfully benefit from caching?" If not, just compute it in the `library/module` function, even if we do that computation a couple more times than we'd ideally like. If we can meaningfully benefit however, then we can make it a (required) argument of the `library/module` function and cache it on the frontend. Importantly, it's either one or the other; not both. We shouldn't make it an optional argument, where if its `None` compute it.
